examples/nextjs-whiteboard-advanced/src/index.tsx

```diff
@@ -438,35 +455,57 @@ function Canvas({
         origin,
         current
       );
-      setPresence({ selection: ids });
+      setMyPresence({ selection: ids });
     },
-    [layers, layerIds, setPresence]
+    [layerIds, layers]
   );
 
-  // TODO: Expose a hook to observe only one key of the others presence to improve performance
-  // For example, multiplayer selection should not be re-render if only a cursor move
-  const others = useOthers();
+  const selections = useOtherIds((other) => other.presence.selection);
 
   /**
    * Create a map layerId to color based on the selection of all the users in the room
    */
   const layerIdsToColorSelection = useMemo(() => {
     const layerIdsToColorSelection: Record<string, string> = {};
 
-    const users = others.toArray();
-    for (const user of users) {
-      const selection = user.presence?.selection;
-      if (selection == null || selection.length === 0) {
-        continue;
-      }
-
+    for (const user of selections) {
+      const { connectionId, data: selection } = user;
       for (const id of selection) {
-        layerIdsToColorSelection[id] = connectionIdToColor(user.connectionId);
+        layerIdsToColorSelection[id] = connectionIdToColor(connectionId);
       }
     }
 
     return layerIdsToColorSelection;
-  }, [others]);
+  }, [selections]);
+
+  const pointerMove = useCallback(
+    (e: React.PointerEvent) => {
+      e.preventDefault();
+      const current = pointerEventToCanvasPoint(e, camera);
+      if (canvasState.mode === CanvasMode.Pressing) {
+        startMultiSelection(current, canvasState.origin);
+      } else if (canvasState.mode === CanvasMode.SelectionNet) {
+        updateSelectionNet(current, canvasState.origin);
+      } else if (canvasState.mode === CanvasMode.Translating) {
+        translateSelectedLayers(current);
+      } else if (canvasState.mode === CanvasMode.Resizing) {
+        resizeSelectedLayer(current);
+      } else if (canvasState.mode === CanvasMode.Pencil) {
+        continueDrawing(current, e);
+      }
+      setPresence({ cursor: current });
+    },
+    [
+      camera,
+      canvasState,
+      continueDrawing,
+      resizeSelectedLayer,
+      setPresence,
+      startMultiSelection,
+      translateSelectedLayers,
+      updateSelectionNet,
+    ]
+  );
```

Comment:
@GuillaumeSalles Wondering if there is a reason why nesting batches throws an error. Could we not safely combine inner batches with the outermost batch silently (making inner batch wrappings a no-op?), or would that be problematic?

I hit this issue when I tried to convert this callback to a `useMutation()` call. Since it will invoke other mutations, this causes nested batching and thus throws an error. That's why I had to make it a `useCallback()`.

Not a super big deal, but it does mean that the outermost `setPresence()` call on L496 is called _after_ the batch, and not _included with the batch_. Not sure if there is a good reason to actively prevent nesting of batches? Because this seems like a useful pattern to support.

Reply:
We throw for nested batches because I never took the time to implement it properly 😬 

It shouldn't be too hard to implement, we just need to be cautious about undo/redo and dispatching. When a batch is finished, we shouldn't send the operations over the wire or add them to the undo/redo stack if there is an outer batch in progress (while being cautious about exceptions coming from the callback).

We can probably squeeze that in 0.18. What do you think?

Reply:
Ah cool! Also I just discovered this ticket: #416 👍  Can take a look tomorrow to see if it's easy to implement. If so, would love to squeeze it in.